### ISSUE TITLE
App update/removal: what to do with data/settings files?

### ISSUE THREAD

**Issue by rigrig** [espruino/BangleApps#243]
Some apps create files on the device, how should update/remove handle those?

As I see it, there are multiple categories of these files:
* "settings": Used to save app settings, typically just one small file with known name
* "data": Used to save app data (gps tracks, heart rates, etc), these can be multiple files (e.g. `.gpsrc1`, `.gpsrc2`, etc), and might actually grow quite large
* "state": Used to save app state between reboots, e.g. the pedometer widget last known steps+date

I make this distinction because losing "settings" or "state" seems to me not as big a deal as losing "data":
You can always just reconfigure the app, but that data is gone forever.
On the other hand, having to reconfigure an app or seeing your step count reset after every update would be annoying.

Some solutions I can think of:
1. Have some way for apps to declare these files in `apps.json`, so the uninstaller can clean them up.
   Possibly this could include a filename prefix, to e.g. clean up all `.gpsrc*` files.
   This would be quite easy to implement, but I'm not too sure about the idea of just deleting all GPS tracks when someone uninstalls the app.
2. Apps should add created files to `<appid>.info` 
   This seems like a lot of hassle to implement in all apps though
3. Implement something similar to `interface.html`(or adapt that) to handle uninstalling. It could just silently clean up settings, or check for data files and ask the user what to do with them.
   This seems like a lot of hassle for all apps that only need to clean up one settings file though.

Maybe I'm being too paranoid, and option 1. would be fine?

**Comment by gfwilliams** [comment]
> Have some way for apps to declare these files in apps.json

I think this is preferable. Maybe as you mentioned above we need separate categories for these kinds of files. Some kind of wildcard/regex might be a good idea as we're moving more towards having file extensions.

The other gotcha is right now, StorageFile and Storage are slightly different and have to be deleted in different ways :(

**Comment by rigrig** [comment]
I mostly thought of these categories as different use cases, not sure if they need actual categorizing.
(Especially not if we are going to treat them all the same anyway)
<br>

> StorageFile and Storage are slightly different and have to be deleted in different ways

Uhm, so what happens if you mess this up? Could it just try both?
```js
storage.erase(filename); storage.open(filename, 'r').erase()
```
<br>

I also just noticed the documentation mentioning an 8-char limit on filenames, that doesn't seem to be true, or does it only apply to storageFiles?

**Comment by gfwilliams** [comment]
> I also just noticed the documentation mentioning an 8-char limit on filenames

Can you link to this? That *was* the case in 2v04, but not in what's used on the bangles. It's now 28.

> Uhm, so what happens if you mess this up? Could it just try both?

Good idea! Yes, that'd work fine. It'd still be better to make it explicit in the JSON though - we only have to worry about it for app-generated files though. The size of stuff being uploaded is known in advance so normal files can be used.

**Comment by rigrig** [comment]
8-char limit: https://banglejs.com/reference#Storage and https://github.com/espruino/BangleApps#what-filenames-are-used

> that'd work fine. It'd still be better to make it explicit in the JSON though

I agree, mostly was a bit worried that getting it wrong might mess up the filesystem.

How about something like
```json
"data": [
  {"name":"myapp.settings.json","type":"file"},
  {"name":"myapp.data","type":"storageFile"},
  {"pattern":"myapp.*.json","type":"file"},
  {"pattern":"myapp.rec?","type":"storageFile"},
]
```
After thinking it over a bit, I feel that simply not uploading them might be best: so only use this for cleanup on uninstall.
Cleaning up "pattern" would have to use `Storage.list()`, and `sanitycheck` should probably complain about too broad patterns.

**Comment by rigrig** [comment]
Actually, maybe it would be a good idea to (also) just hardcode in the removal of `<appid>.settings.json`. No need to clutter up pretty much any app with settings for that.

**Comment by gfwilliams** [comment]
Just wanted to add, while I'd like to get this fixed, I think the various settings files are likely to be pretty small and are mostly harmless - so IMO this isn't a super high priority thing to fix. I think the fallout from a cludgy fix is likely to be worse than if we wait until we can get a really nice maintainable way of sorting it out.

**Comment by rigrig** [comment]
Fixed with #281
